internal/jsonwall/jsonwall.go

```diff
@@ -0,0 +1,307 @@
+	Schema string
+}
+
+// NewDBWriter returns a database writer that can assemble new databases.
+func NewDBWriter(options *DBWriterOptions) *DBWriter {
+	if options == nil {
+		options = &DBWriterOptions{}
+	}
+	return &DBWriter{options: options}
+}
+
+// Add encodes the provided value as a JSON object and includes the resulting
+// data into the database being created.
+func (dbw *DBWriter) Add(value any) error {
+	data, err := json.Marshal(value)
+	if err != nil {
+		return err
+	}
+	if len(data) == 0 || data[0] != '{' {
+		return fmt.Errorf("invalid database value: %#v", value)
+	}
+	dbw.entries = append(dbw.entries, data)
+	return nil
+}
+
+type sortableEntries [][]byte
+
+func (e sortableEntries) Less(i, j int) bool { return bytes.Compare(e[i], e[j]) < 0 }
+func (e sortableEntries) Swap(i, j int)      { e[i], e[j] = e[j], e[i] }
+func (e sortableEntries) Len() int           { return len(e) }
+
+type jsonwallHeader struct {
+	Version string `json:"jsonwall"`
+	Schema  string `json:"schema,omitempty"`
+	Count   int    `json:"count"`
+}
+
+const jsonwallVersion = "1.0"
+
+func (dbw *DBWriter) writeHeader(w io.Writer, count int) (int, error) {
+	data, err := json.Marshal(&jsonwallHeader{
+		Version: jsonwallVersion,
+		Schema:  dbw.options.Schema,
+		Count:   count,
+	})
+	if err != nil {
+		return 0, fmt.Errorf("internal error: cannot marshal database header")
+	}
+	return w.Write(append(data, '\n'))
+}
+
+// WriteTo assembles the current database state and writes it to w.
+func (dbw *DBWriter) WriteTo(w io.Writer) (n int64, err error) {
+	m, err := dbw.writeHeader(w, len(dbw.entries)+1)
+	n += int64(m)
+	sort.Sort(sortableEntries(dbw.entries))
+	if err != nil {
+		return n, err
+	}
+	for _, entry := range dbw.entries {
+		m, err := w.Write(entry)
+		n += int64(m)
+		if err == nil {
+			m, err = w.Write([]byte{'\n'})
+			n += int64(m)
+		}
+		if err != nil {
+			return n, err
+		}
+	}
+	return n, nil
+}
+
+// ReadDB reads into memory the database from the provided r.
+func ReadDB(r io.Reader) (*DB, error) {
+	data, err := ioutil.ReadAll(r)
+	if err != nil {
+		return nil, err
+	}
+	record := 0
+	for i := range data {
+		if data[i] == '\n' {
+			record = i + 1
+			break
+		}
+	}
+	var header jsonwallHeader
+	err = json.Unmarshal(data[:record], &header)
+	if err != nil {
+		return nil, fmt.Errorf("invalid database content")
+	}
+	if !strings.HasPrefix(header.Version, jsonwallVersion[:strings.Index(jsonwallVersion, ".")+1]) {
+		return nil, fmt.Errorf("unsupported database format: %q", header.Version)
+	}
+	if header.Count > len(data)/8 {
+		// The header helps pre-allocating an index of the right size,
+		// but it could trivially be abused to cause an OOM situation.
+		header.Count = 0
+	}
+	db := &DB{schema: header.Schema, data: data}
+	db.index = make([]int, 0, header.Count)
+	for i := range data {
+		if data[i] == '\n' {
+			db.index = append(db.index, i+1)
+		}
+	}
+	db.count = len(db.index) - 1 // Last item is end of buffer.
```

Comment:
Yeah, it's definitely not a format friendly to manual edits, but we may as well handle the trivial cases of a missing newline or extra newlines, since we can probably do that cheaply outside of the fast path (that particular loop will ingest a lot of data).